python/pyspark/find_spark_home.py

```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python
+
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# This script attempt to determine the correct setting for SPARK_HOME given
+# that Spark may have been installed on the system with pip.
+
+from __future__ import print_function
+import os
+import sys
+
+
+def _find_spark_home():
+    """Find the SPARK_HOME."""
+    # If the enviroment has SPARK_HOME set trust it.
+    if "SPARK_HOME" in os.environ:
+        return os.environ["SPARK_HOME"]
+
+    def is_spark_home(path):
+        """Takes a path and returns true if the provided path could be a reasonable SPARK_HOME"""
+        return (os.path.isfile(os.path.join(path, "bin/spark-submit")) and
+                (os.path.isdir(os.path.join(path, "jars"))))
+
+    paths = ["../", os.path.join(os.path.dirname(sys.argv[0]), "../")]
```

Comment:
Couple of comments here:

1. A better way to get a directory relative to the current file is to have something like this at the top of the file and refer to it as necessary:

   ```
   THIS_DIR = os.path.dirname(os.path.realpath(__file__))
   ```

2. The correct way to go up one directory is to just call `dirname()` again. `os.path.dirname(..., '../')` will just append `'../'` to the end of the path, which may not work as expected later on.

  So I think you're looking for `THIS_DIR` and `os.dirname(THIS_DIR)`.